cdap-common/src/main/java/io/cdap/cdap/common/discovery/preview/PreviewDiscoveryService.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package io.cdap.cdap.common.discovery.preview;
+
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import io.cdap.cdap.common.guice.preview.PreviewDiscoveryRuntimeModule;
+import io.netty.util.internal.ConcurrentSet;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.DiscoveryService;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.discovery.InMemoryDiscoveryService;
+import org.apache.twill.discovery.ServiceDiscovered;
+
+import java.util.Set;
+
+/**
+ * Discovery service that delegates either to in memory discovery service(local) or shared (actual) discovery service.
+ */
+public class PreviewDiscoveryService implements DiscoveryService, DiscoveryServiceClient {
+  private final DiscoveryServiceClient actual;
+  private final InMemoryDiscoveryService local;
+  private final Set<String> registeredLocalServices;
+
+  @Inject
+  public PreviewDiscoveryService(
+    @Named(PreviewDiscoveryRuntimeModule.ACTUAL_DISCOVERY_CLIENT) DiscoveryServiceClient actual,
+    InMemoryDiscoveryService local) {
```

Comment:
Why the `InMemoryDiscoveryService` needs to be injected? You can just `new` it in this constructor, right?

Reply:
Fixed to instantiate it in the constructor